File: Leetcode/C++/64. Minimum Path Sum/solution.cpp
// https://leetcode.com/problems/minimum-path-sum/description/
// Time complexity: O(N * M), N = grid.length, M = grid[0].length

#include <vector>

using namespace std;

class Solution
{
public:
    int minPathSum(vector<vector<int>> &grid)
    {
        int lenX = grid[0].size(), lenY = grid.size();

        for (int y = 0; y < lenY; y++)
        {
            for (int x = 0; x < lenX; x++)
            {
                if (y && x)
                {
                    grid[y][x] += min(grid[y - 1][x], grid[y][x - 1]);
                }
                else if (y) // If x == 0 move down
                {
                    grid[y][x] += grid[y - 1][x];
                }
                else if (x) // If y == 0 move right
                {
                    grid[y][x] += grid[y][x - 1];
                }
            }
        }
        return grid[lenY - 1][lenX - 1];
    }
};

File: Leetcode/Python3/64. Minimum Path Sum/solution.py
# https://leetcode.com/problems/minimum-path-sum/description/
# Time complexity: O(N * M), N = grid.length, M = grid[0].length


from collections import deque
from typing import List


class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        len_x, len_y = len(grid[0]), len(grid)
        dp = [[float("inf") for _ in range(len_x)] for _ in range(len_y)]
        # Can move only right, down
        dx = [0, 1]
        dy = [1, 0]

        dp[0][0] = grid[0][0]   # Starting point
        queue = deque([(0, 0)])
        while queue:    # BFS
            x, y = queue.popleft()
            for idx in range(2):
                nx, ny = x + dx[idx], y + dy[idx]
                if 0 <= nx < len_x and 0 <= ny < len_y and dp[ny][nx] > dp[y][x] + grid[ny][nx]:
                    dp[ny][nx] = dp[y][x] + grid[ny][nx]
                    queue.append((nx, ny))
        print(dp)
        return dp[-1][-1]
